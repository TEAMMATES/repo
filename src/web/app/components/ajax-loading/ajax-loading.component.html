<div class="loading-container">
  <div class="spinner-border spinner-border-sm align-middle text-{{useBlueSpinner ? 'primary' : color}}" role="status"></div>
</div>
