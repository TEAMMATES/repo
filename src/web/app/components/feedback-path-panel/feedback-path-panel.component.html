<div class="background-color-light-green margin-top-15px padding-15px">
  <div>
    <b class="feedback-path-title">Feedback Path</b> (Who is giving feedback about whom?)
  </div>
</div>