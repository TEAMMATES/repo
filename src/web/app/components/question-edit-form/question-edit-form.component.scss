.question-number-select {
  height: 32px;
  width: auto;
  display: inline;
  margin-left: 6px;
  margin-right: 5px;
}

.background-color-light-blue {
  background-color: #EAEFF5;
}

.background-color-warning {
  background-color: #FCF8E3;
}

.background-color-white {
  background-color: white;
}

.background-color-light-green {
  background-color: #E0EEEE;
}

.feedback-path-title {
  color: #2AABD2;
}

.dropdown-item {
  cursor: pointer;
}

.visibility-title {
  color: #2AABD2;
}

.visibility-message > ul {
  margin-top: 5px;
  padding: 5px 25px;
  margin-bottom: 0;
}

.visibility-message > ul > li {
  cursor: default;
}

.white-space-normal {
  white-space: normal;
}

.padding-15px {
  padding: 15px;
}

.question-basic-info {
  padding-left: 15px;
  font-weight: bold;
}

.padding-left-15px {
  padding-left: 15px;
}

.margin-top-15px {
  margin-top: 15px;
}

.collapse-caret {
  margin-right: 10px;
}

.collapse-qn-description {
  font-weight: 300;
}
