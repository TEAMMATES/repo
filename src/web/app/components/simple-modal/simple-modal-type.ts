/**
 * Represents the type of  the modal to be displayed.
 */
export enum SimpleModalType {
  /**
   * Warning modal.
   */
  WARNING,
  /**
   * Danger modal.
   */
  DANGER,
  /**
   * Info modal.
   */
  INFO,
  /**
   * Primary modal.
   */
  PRIMARY,
  /**
   * Neutral modal.
   */
  NEUTRAL,
}
