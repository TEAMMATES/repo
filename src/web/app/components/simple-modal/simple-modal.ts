import { TemplateRef } from '@angular/core';

/**
 * Represents a button for the modal.
 */
export interface SimpleModalButton {
  text: string;
  classes?: string;
  action?: Function;
}

/**
 * The standard cancel button for modal.
 */
export const standardCancelButton: (text?: string) => SimpleModalButton = (text?: string) => {
  return {
    text: text || 'Cancel',
    classes: 'btn-light modal-btn-cancel',
  };
};

/**
 * The standard confirm button for modal.
 */
export const standardConfirmButton: (type: string, action: Function, text?: string) => SimpleModalButton =
    (type: string, action: Function, text?: string) => {
      return {
        action,
        text: text || 'Yes',
        classes: `${type} modal-btn-ok`,
      };
    };

/**
 * Parameters for modal.
 */
export interface SimpleModalOptions {
  type: SimpleModalType;
  header: string | TemplateRef<any>;
  content: string | TemplateRef<any>;
  context?: Record<string, any>;
  buttons?: SimpleModalButton[];
  options?: Record<string, string | boolean>;
}

/**
 * Represents the type of the modal to be displayed.
 */
export enum SimpleModalType {
  /**
   * Warning modal.
   */
  WARNING,
  /**
   * Danger modal.
   */
  DANGER,
  /**
   * Info modal.
   */
  INFO,
  /**
   * Neutral modal.
   */
  NEUTRAL,
}
