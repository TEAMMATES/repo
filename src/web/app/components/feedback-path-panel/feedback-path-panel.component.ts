import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import {
  FeedbackParticipantType,
  FeedbackQuestionType,
  FeedbackTextQuestionDetails,
  NumberOfEntitiesToGiveFeedbackToSetting,
} from '../../../types/api-output';
import { QuestionEditFormModel } from '../question-edit-form/question-edit-form-model';

/**
 * Displaying the feedback path panel.
 */
@Component({
  selector: 'tm-feedback-path-panel',
  templateUrl: './feedback-path-panel.component.html',
  styleUrls: ['./feedback-path-panel.component.scss'],
})
export class FeedbackPathPanelComponent implements OnInit {

  // enum
  FeedbackParticipantType: typeof FeedbackParticipantType = FeedbackParticipantType;
  FeedbackQuestionType: typeof FeedbackQuestionType = FeedbackQuestionType;
  NumberOfEntitiesToGiveFeedbackToSetting: typeof NumberOfEntitiesToGiveFeedbackToSetting =
      NumberOfEntitiesToGiveFeedbackToSetting;

  @Input()
  model: QuestionEditFormModel = {
    feedbackQuestionId: '',

    questionNumber: 0,
    questionBrief: '',
    questionDescription: '',

    isQuestionHasResponses: false,

    questionType: FeedbackQuestionType.TEXT,
    questionDetails: {
      questionType: FeedbackQuestionType.TEXT,
      questionText: '',
    } as FeedbackTextQuestionDetails,

    giverType: FeedbackParticipantType.STUDENTS,
    recipientType: FeedbackParticipantType.OWN_TEAM_MEMBERS,

    numberOfEntitiesToGiveFeedbackToSetting: NumberOfEntitiesToGiveFeedbackToSetting.UNLIMITED,
    customNumberOfEntitiesToGiveFeedbackTo: 1,

    showResponsesTo: [],
    showGiverNameTo: [],
    showRecipientNameTo: [],

    commonVisibilitySettingName: '',

    isUsingOtherFeedbackPath: false,
    isUsingOtherVisibilitySetting: false,
    isDeleting: false,
    isDuplicating: false,
    isEditable: false,
    isSaving: false,
    isCollapsed: false,
    isVisibilityChanged: false,
    isFeedbackPathChanged: false,
    isQuestionDetailsChanged: false,
  };

  @Input()
  commonFeedbackPaths: Map<FeedbackParticipantType, FeedbackParticipantType[]> = new Map();

  @Input()
  allowedFeedbackPaths: Map<FeedbackParticipantType, FeedbackParticipantType[]> = new Map();

  @Output()
  changeGiverRecipientTypeEvent: EventEmitter<{ giverType: FeedbackParticipantType,
    recipientType: FeedbackParticipantType }> = new EventEmitter<{ giverType: FeedbackParticipantType,
      recipientType: FeedbackParticipantType }>();

  @Output()
  triggerModelChangeEvent: EventEmitter<{ field: keyof QuestionEditFormModel,
    data: QuestionEditFormModel[keyof QuestionEditFormModel] }> = new EventEmitter<{ field: keyof QuestionEditFormModel,
      data: QuestionEditFormModel[keyof QuestionEditFormModel] }>();

  constructor() { }

  ngOnInit(): void {
  }

  /**
   * Handles the changing of the {@code giverType} and {@code recipientType} and reset the visibility settings.
   */
  changeGiverRecipientTypeHandler(giverType: FeedbackParticipantType, recipientType: FeedbackParticipantType): void {
    this.changeGiverRecipientTypeEvent.emit({ giverType, recipientType });
  }

  /**
   * Handles triggering of the change of the model for the form.
   */
  triggerModelChangeHandler(field: keyof QuestionEditFormModel,
    data: QuestionEditFormModel[keyof QuestionEditFormModel]): void {
    this.triggerModelChangeEvent.emit({ field, data });
  }
}
