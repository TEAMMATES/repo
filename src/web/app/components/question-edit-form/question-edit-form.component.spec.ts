import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { FormsModule } from '@angular/forms';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import SpyInstance = jest.SpyInstance;

import { SimpleModalService } from '../../../services/simple-modal.service';
import { createMockNgbModalRef } from '../../../test-helpers/mock-ngb-modal-ref';
import { mockTinyMceUuid } from '../../../test-helpers/mock-tinymce-uuid';
import { FeedbackParticipantType, NumberOfEntitiesToGiveFeedbackToSetting } from '../../../types/api-output';
import { FeedbackVisibilityType } from '../../../types/api-request';
import {
  EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES,
} from '../../pages-help/instructor-help-page/instructor-help-questions-section/instructor-help-questions-data';
import { AjaxLoadingModule } from '../ajax-loading/ajax-loading.module';
import { FeedbackPathPanelModule } from '../feedback-path-panel/feedback-path-panel.module';
import { PanelChevronModule } from '../panel-chevron/panel-chevron.module';
import {
  QuestionEditBriefDescriptionFormModule,
} from '../question-edit-brief-description-form/question-edit-brief-description-form.module';
import {
  QuestionEditDetailsFormModule,
} from '../question-types/question-edit-details-form/question-edit-details-form.module';
import { SimpleModalType } from '../simple-modal/simple-modal-type';
import { TeammatesCommonModule } from '../teammates-common/teammates-common.module';
import { VisibilityMessagesModule } from '../visibility-messages/visibility-messages.module';
import { VisibilityPanelModule } from '../visibility-panel/visibility-panel.module';
import { QuestionEditFormModel } from './question-edit-form-model';
import { QuestionEditFormComponent } from './question-edit-form.component';

describe('QuestionEditFormComponent', () => {
  let component: QuestionEditFormComponent;
  let fixture: ComponentFixture<QuestionEditFormComponent>;
  let simpleModalService: SimpleModalService;

  /**
   * Helper function to check if a model is equal to
   * EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES.
   */
  const verifyEqualsToExampleQuestionModel: (model: QuestionEditFormModel) => void =
  (model: QuestionEditFormModel): void => {
    expect(model).toStrictEqual(EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES);
    expect(model.isUsingOtherFeedbackPath).toBeFalsy();
    expect(model.commonVisibilitySettingName).toBeTruthy();
    expect(model.isUsingOtherVisibilitySetting).toBeFalsy();
    expect(component.isCustomFeedbackVisibilitySettingAllowed).toBeTruthy();
  };

  mockTinyMceUuid();

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [
        QuestionEditFormComponent,
      ],
      imports: [
        HttpClientTestingModule,
        FormsModule,
        TeammatesCommonModule,
        AjaxLoadingModule,
        QuestionEditBriefDescriptionFormModule,
        QuestionEditDetailsFormModule,
        NgbModule,
        FeedbackPathPanelModule,
        VisibilityMessagesModule,
        VisibilityPanelModule,
        BrowserAnimationsModule,
        PanelChevronModule,
      ],
      providers: [
        SimpleModalService,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(QuestionEditFormComponent);
    simpleModalService = TestBed.inject(SimpleModalService);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('set up with EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES', () => {
    component.formModel = EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES;
    const model: QuestionEditFormModel = component.model;
    verifyEqualsToExampleQuestionModel(model);

    const unsavedModel: QuestionEditFormModel = component.model;
    verifyEqualsToExampleQuestionModel(unsavedModel);
    expect(unsavedModel).toStrictEqual(model);
  });

  it('triggerUnsavedModelChange with EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES', () => {
    component.formModel = EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES;

    const testStr: string = 'Hello World';
    component.triggerUnsavedModelChange('questionNumber', 2);
    component.triggerUnsavedModelChange('questionBrief', testStr);
    component.triggerUnsavedModelChange('questionDescription', testStr);
    component.triggerUnsavedModelChange('isQuestionHasResponses', true);

    const model: QuestionEditFormModel = component.model;
    verifyEqualsToExampleQuestionModel(model);

    const unsavedModel: QuestionEditFormModel = component.unsavedModel;
    expect(unsavedModel.questionNumber).toBe(2);
    expect(unsavedModel.questionBrief).toBe(testStr);
    expect(unsavedModel.questionDescription).toBe(testStr);
    expect(unsavedModel.isQuestionHasResponses).toBeTruthy();
  });

  it('should snap with default view', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('triggerUnsavedModelChangeBatch with EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES', () => {
    component.formModel = EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES;

    const testObj: Partial<QuestionEditFormModel> = {
      giverType: FeedbackParticipantType.STUDENTS,
      recipientType: FeedbackParticipantType.INSTRUCTORS,
      isUsingOtherFeedbackPath: false,
      numberOfEntitiesToGiveFeedbackToSetting:
      NumberOfEntitiesToGiveFeedbackToSetting.UNLIMITED,
      customNumberOfEntitiesToGiveFeedbackTo: undefined,

    };
    component.triggerUnsavedModelChangeBatch(testObj);

    const model: QuestionEditFormModel = component.model;
    verifyEqualsToExampleQuestionModel(model);

    const unsavedModel: QuestionEditFormModel = component.unsavedModel;
    expect(unsavedModel.giverType).toBe(testObj.giverType);
    expect(unsavedModel.recipientType).toBe(testObj.recipientType);
    expect(unsavedModel.isUsingOtherFeedbackPath).toBeFalsy();
    expect(unsavedModel.numberOfEntitiesToGiveFeedbackToSetting)
    .toBe(testObj.numberOfEntitiesToGiveFeedbackToSetting);
    expect(unsavedModel.customNumberOfEntitiesToGiveFeedbackTo)
    .toBe(testObj.customNumberOfEntitiesToGiveFeedbackTo);
  });

  it('discardChangesHandler with EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES', async () => {
    component.formModel = EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES;

    component.triggerUnsavedModelChange('isUsingOtherFeedbackPath', true);
    component.triggerUnsavedModelChange('showResponsesTo', [FeedbackVisibilityType.RECIPIENT]);
    component.triggerUnsavedModelChange('showGiverNameTo', [FeedbackVisibilityType.RECIPIENT]);
    component.triggerUnsavedModelChange('showRecipientNameTo', [FeedbackVisibilityType.RECIPIENT]);
    component.triggerUnsavedModelChange('giverType', FeedbackParticipantType.STUDENTS_EXCLUDING_SELF);

    const promise: Promise<void> = Promise.resolve();
    const modalSpy: SpyInstance = jest.spyOn(simpleModalService, 'openConfirmationModal')
        .mockReturnValue(createMockNgbModalRef({}, promise));
    component.discardChangesHandler(false);
    await promise;
    expect(modalSpy).toHaveBeenCalledTimes(1);
    expect(modalSpy).toHaveBeenLastCalledWith('Discard unsaved edits?', SimpleModalType.WARNING,
    'Warning: Any unsaved changes will be lost',
    { cancelMessage: 'No, go back to editing' });
  });

  it('discardChangesHandler with EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES where changes reverted', async () => {
    component.formModel = EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES;

    component.triggerUnsavedModelChange('giverType', FeedbackParticipantType.STUDENTS_EXCLUDING_SELF);
    component.triggerUnsavedModelChange('giverType', FeedbackParticipantType.STUDENTS);

    const promise: Promise<void> = Promise.resolve();
    const modalSpy: SpyInstance = jest.spyOn(simpleModalService, 'openConfirmationModal')
        .mockReturnValue(createMockNgbModalRef({}, promise));
    component.discardChangesHandler(false);
    await promise;
    expect(modalSpy).toHaveBeenCalledTimes(0);
  });

  it('saveQuestionHandler with EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES where feedback path changed', () => {
    component.formModel = EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES;

    component.triggerUnsavedModelChange('isQuestionHasResponses', true);
    component.triggerUnsavedModelChange('recipientType', FeedbackParticipantType.STUDENTS_EXCLUDING_SELF);

    const promise: Promise<void> = Promise.resolve();
    const modalSpy: SpyInstance = jest.spyOn(simpleModalService, 'openConfirmationModal')
        .mockReturnValue(createMockNgbModalRef({}, promise));
    component.saveQuestionHandler();
    expect(modalSpy).toHaveBeenCalledTimes(1);

    const modalContent: string = `
            <p>You seem to have changed the feedback path settings of this question. Please note that changing the
            feedback path will cause <b>all existing responses to be deleted.</b> Proceed?</p>
        `;
    expect(modalSpy).toHaveBeenCalledWith('Save the question?', SimpleModalType.DANGER, modalContent);
  });

  it('saveQuestionHandler with EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES where feedback path reverted', () => {
    component.formModel = EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES;

    component.triggerUnsavedModelChange('isQuestionHasResponses', true);
    component.triggerUnsavedModelChange('recipientType', FeedbackParticipantType.STUDENTS_EXCLUDING_SELF);
    component.triggerUnsavedModelChange('recipientType', FeedbackParticipantType.OWN_TEAM_MEMBERS);

    const promise: Promise<void> = Promise.resolve();
    const modalSpy: SpyInstance = jest.spyOn(simpleModalService, 'openConfirmationModal')
        .mockReturnValue(createMockNgbModalRef({}, promise));
    component.saveQuestionHandler();
    expect(modalSpy).toHaveBeenCalledTimes(0);
  });

  it('saveQuestionHandler with EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES where question details changed', () => {
    component.formModel = EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES;

    component.triggerUnsavedModelChange('isQuestionHasResponses', true);
    component.triggerUnsavedModelChange('questionDescription', 'A new description for this question');

    const promise: Promise<void> = Promise.resolve();
    const modalSpy: SpyInstance = jest.spyOn(simpleModalService, 'openConfirmationModal')
        .mockReturnValue(createMockNgbModalRef({}, promise));
    component.saveQuestionHandler();

    expect(modalSpy).toHaveBeenCalledTimes(1);

    const modalContent: string = `
            <p>Editing question settings in a way that potentially affects the validity of existing responses <b> may
            cause all the existing responses for this question to be deleted.</b> Proceed?</p>
        `;
    expect(modalSpy).toHaveBeenCalledWith('Save the question?', SimpleModalType.DANGER, modalContent);
  });

  it('saveQuestionHandler with EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES where question details reverted', () => {
    component.formModel = EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES;

    component.triggerUnsavedModelChange('isQuestionHasResponses', true);
    component.triggerUnsavedModelChange('questionDescription', 'A new description for this question');
    component.triggerUnsavedModelChange('questionDescription', '');

    const promise: Promise<void> = Promise.resolve();
    const modalSpy: SpyInstance = jest.spyOn(simpleModalService, 'openConfirmationModal')
        .mockReturnValue(createMockNgbModalRef({}, promise));
    component.saveQuestionHandler();
    expect(modalSpy).toHaveBeenCalledTimes(0);
  });

  it('saveQuestionHandler with EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES where visibility changed', () => {
    component.formModel = EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES;

    component.triggerUnsavedModelChange('isQuestionHasResponses', true);
    component.triggerUnsavedModelChange('showRecipientNameTo',
    [FeedbackVisibilityType.INSTRUCTORS, FeedbackVisibilityType.RECIPIENT, FeedbackVisibilityType.STUDENTS]);

    const promise: Promise<void> = Promise.resolve();
    const modalSpy: SpyInstance = jest.spyOn(simpleModalService, 'openConfirmationModal')
        .mockReturnValue(createMockNgbModalRef({}, promise));
    component.saveQuestionHandler();
    expect(modalSpy).toHaveBeenCalledTimes(1);

    const modalContent: string = `
            <p>You seem to have changed the visibility settings of this question. Please note that <b>the existing
            responses will remain but their visibility will be changed as per the new visibility settings.</b>
            Proceed?</p>
        `;
    expect(modalSpy).toHaveBeenCalledWith('Save the question?', SimpleModalType.WARNING, modalContent);
  });

  it('saveQuestionHandler with EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES where visibility reverted', () => {
    component.formModel = EXAMPLE_ESSAY_QUESTION_MODEL_WITH_RESPONSES;

    component.triggerUnsavedModelChange('isQuestionHasResponses', true);
    component.triggerUnsavedModelChange('showRecipientNameTo',
    [FeedbackVisibilityType.INSTRUCTORS, FeedbackVisibilityType.RECIPIENT, FeedbackVisibilityType.STUDENTS]);
    component.triggerUnsavedModelChange('showRecipientNameTo',
    [FeedbackVisibilityType.INSTRUCTORS, FeedbackVisibilityType.RECIPIENT]);

    const promise: Promise<void> = Promise.resolve();
    const modalSpy: SpyInstance = jest.spyOn(simpleModalService, 'openConfirmationModal')
        .mockReturnValue(createMockNgbModalRef({}, promise));
    component.saveQuestionHandler();
    expect(modalSpy).toHaveBeenCalledTimes(0);
  });
});
