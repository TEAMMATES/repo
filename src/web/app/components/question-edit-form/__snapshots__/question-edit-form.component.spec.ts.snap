// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionEditFormComponent should snap with default view 1`] = `
<tm-question-edit-form
  FeedbackParticipantType={[Function Object]}
  FeedbackQuestionType={[Function Object]}
  FeedbackVisibilityType={[Function Object]}
  NumberOfEntitiesToGiveFeedbackToSetting={[Function Object]}
  QuestionEditFormMode={[Function Object]}
  VisibilityControl={[Function Object]}
  allowedFeedbackPaths={[Function Map]}
  commonFeedbackPaths={[Function Map]}
  commonFeedbackVisibilitySettings={[Function Array]}
  createNewQuestionEvent={[Function EventEmitter]}
  deleteCurrentQuestionEvent={[Function EventEmitter]}
  discardExistingQuestionChangesEvent={[Function EventEmitter]}
  discardNewQuestionEvent={[Function EventEmitter]}
  duplicateCurrentQuestionEvent={[Function EventEmitter]}
  feedbackQuestionsService={[Function FeedbackQuestionsService]}
  formMode={[Function Number]}
  formModelChange={[Function EventEmitter]}
  isDisplayOnly="false"
  modalService={[Function NgbModal]}
  model={[Function Object]}
  numOfQuestions="0"
  saveExistingQuestionEvent={[Function EventEmitter]}
  visibilityStateMachine={[Function VisibilityStateMachine]}
>
  <div
    class="card"
    xmlns="http://www.w3.org/1999/html"
  >
    <div
      class="card-header text-white question-header"
    >
      <div
        class="row align-items-center"
      >
        <div
          class="col-5"
        >
          <strong>
            Question 
            
            <span>
              0
            </span>
          </strong>
          
           Essay question 
        </div>
        
        <div
          class="col-7 text-right"
        >
          
          <div>
            
            <button
              class="btn btn-primary btn-sm"
              ng-reflect-ngb-tooltip="Edit the existing question. Do"
              ngbtooltip="Edit the existing question. Do remember to save the changes before moving on to editing another question."
              type="button"
            >
              <i
                class="fas fa-pencil-alt"
              />
               Edit
            </button>
            
            
            
            
            <button
              class="btn btn-primary btn-sm"
              type="button"
            >
              <i
                class="fas fa-trash"
              />
               Delete
            </button>
            
            <button
              class="btn btn-primary btn-sm"
              container="body"
              ng-reflect-container="body"
              ng-reflect-ngb-tooltip="Make a copy of the existing qu"
              ngbtooltip="Make a copy of the existing question and add to the current feedback session."
              type="button"
            >
              <i
                class="far fa-copy"
              />
               Duplicate
            </button>
          </div>
          
        </div>
      </div>
    </div>
    <div
      class="card-body"
    >
      <div
        class="background-color-light-blue"
      >
        <div
          class="row"
        >
          <div
            class="col-12"
          >
            <div
              class="row padding-15px align-items-center"
            >
              <div
                class="col-2"
              >
                <div
                  class="question-basic-info"
                >
                  Question
                </div>
              </div>
              <div
                class="col-10"
              >
                <textarea
                  class="form-control ng-untouched ng-pristine ng-valid"
                  disabled=""
                  ng-reflect-is-disabled="true"
                  ng-reflect-model=""
                  ng-reflect-ngb-tooltip="Please enter the question for "
                  ngbtooltip="Please enter the question for users to give feedback about. e.g. What is the biggest weakness of the presented product?"
                  placeholder="A concise version of the question e.g. \\"How well did the team member communicate?\\""
                />
              </div>
            </div>
            <div
              class="row padding-15px"
            >
              <div
                class="col-2"
              >
                <div
                  class="question-basic-info"
                >
                  [Optional]
                  <br />
                  Description
                </div>
              </div>
              <div
                class="col-10"
              >
                <tm-rich-text-editor
                  ng-reflect-is-disabled="true"
                  ng-reflect-ngb-tooltip="Please enter the description o"
                  ng-reflect-placeholder-text="More details about the questio"
                  ng-reflect-rich-text=""
                  ngbtooltip="Please enter the description of the question."
                  placeholdertext="More details about the question e.g. \\"In answering the question, do consider communications made informally within the team, and formal communications with the instructors and tutors.\\""
                >
                  <editor
                    _nghost-a-c1=""
                    class="editor editor-disabled ng-untouched ng-pristine ng-valid"
                    ng-reflect-disabled="true"
                    ng-reflect-init="[object Object]"
                    ng-reflect-inline="true"
                    ng-reflect-is-disabled="true"
                    ng-reflect-klass="editor"
                    ng-reflect-model=""
                    ng-reflect-ng-class="[object Object]"
                    ng-reflect-ng-style="[object Object]"
                    placeholder="More details about the question e.g. \\"In answering the question, do consider communications made informally within the team, and formal communications with the instructors and tutors.\\""
                    style="min-height: 120px;"
                  >
                    
                    <div
                      id="tiny-angular_031577836800000"
                    />
                  </editor>
                </tm-rich-text-editor>
              </div>
            </div>
            <div
              class="padding-15px"
            >
              
              <tm-text-question-edit-details-form
                ng-reflect-details="[object Object]"
                ng-reflect-is-editable="false"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-12 question-recommended-length"
                  >
                     [Optional] 
                    <span
                      ng-reflect-ngb-tooltip="The recommended length is show"
                      ngbtooltip="The recommended length is shown to the respondent but not enforced"
                    >
                       Recommended length 
                    </span>
                     for the response: 
                    <input
                      class="form-control ng-untouched ng-pristine ng-valid"
                      disabled=""
                      min="0"
                      ng-reflect-is-disabled="true"
                      ng-reflect-model="0"
                      type="number"
                    />
                     words 
                  </div>
                </div>
              </tm-text-question-edit-details-form>
              
              
              
              
              
              
              
              
              
            </div>
          </div>
        </div>
      </div>
      <div
        class="background-color-light-green margin-top-15px padding-15px"
      >
        <div>
          <b
            class="feedback-path-title"
          >
            Feedback Path
          </b>
           (Who is giving feedback about whom?) 
        </div>
        <div
          class="margin-top-15px dropdown"
          ngbdropdown=""
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="btn btn-light dropdown-toggle"
            disabled=""
            ngbdropdowntoggle=""
          >
            
            <span>
               Students in this course will give feedback on 
              <i
                class="fas fa-arrow-right"
              />
               Other students in the course 
            </span>
            
          </button>
          <ul
            class="dropdown-menu"
            ngbdropdownmenu=""
            x-placement="bottom"
          >
            <li
              class="dropdown-header"
            >
              Common feedback path combinations
            </li>
            
            <li
              class="dropdown-divider"
            />
            <li
              class="dropdown-item"
            >
              Custom feedback path...
            </li>
          </ul>
        </div>
        
        
        <div
          class="row margin-top-15px align-items-center"
        >
          <div
            class="offset-1 col-4 text-right"
          >
             The maximum number of students each respondent should give feedback to: 
          </div>
          <div
            class="col-2"
          >
            <div
              class="input-group"
            >
              <div
                class="input-group-prepend"
              >
                <div
                  class="input-group-text"
                >
                  <input
                    class="ng-untouched ng-pristine ng-valid"
                    disabled=""
                    ng-reflect-is-disabled="true"
                    ng-reflect-model="UNLIMITED"
                    ng-reflect-name="numOfRecipients"
                    ng-reflect-value="CUSTOM"
                    type="radio"
                  />
                </div>
              </div>
              <input
                class="form-control ng-untouched ng-pristine ng-valid"
                disabled=""
                min="1"
                ng-reflect-is-disabled="true"
                ng-reflect-model="1"
                type="number"
              />
            </div>
          </div>
          <div
            class="col-2"
          >
            <div
              class="form-check"
            >
              <label
                class="form-check-label"
              >
                <input
                  class="form-check-input ng-untouched ng-pristine ng-valid"
                  disabled=""
                  ng-reflect-is-disabled="true"
                  ng-reflect-model="UNLIMITED"
                  ng-reflect-name="numOfRecipients"
                  ng-reflect-value="UNLIMITED"
                  type="radio"
                />
                 Unlimited 
              </label>
            </div>
          </div>
        </div>
      </div>
      <div
        class="margin-top-15px padding-15px background-color-light-green"
        ng-reflect-klass="margin-top-15px padding-15px"
        ng-reflect-ng-class="[object Object]"
      >
        <div>
          
          <b
            class="visibility-title"
            ng-reflect-ng-class="[object Object]"
          >
            Visibility
          </b>
           (Who can see the responses?) 
        </div>
        <div
          class="margin-top-15px dropdown"
          ngbdropdown=""
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="btn btn-light dropdown-toggle"
            disabled=""
            ngbdropdowntoggle=""
          >
            
            <span>
                
            </span>
            
          </button>
          <ul
            class="dropdown-menu"
            ngbdropdownmenu=""
            x-placement="bottom"
          >
            <li
              class="dropdown-header"
            >
              Common visibility options
            </li>
            
            
            <li
              class="dropdown-divider"
            />
            
            <li
              class="dropdown-item"
            >
              Custom visibility options...
            </li>
          </ul>
        </div>
        
        <div
          class="margin-top-15px visibility-message"
        >
           This is the visibility hint as seen by the feedback giver: 
          <ul
            class="text-muted background-color-warning"
          >
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            <li>
              No-one can see your responses
            </li>
          </ul>
        </div>
      </div>
      
      <div
        class="row margin-top-15px"
      >
        <div
          class="col-12 text-right"
        >
          <button
            class="btn btn-primary"
            disabled=""
          >
            
            
            <span>
              Save Changes
            </span>
            
          </button>
        </div>
      </div>
    </div>
  </div>
</tm-question-edit-form>
`;
